internal/controllers/machine/machine_controller.go

```diff
@@ -488,6 +499,74 @@ func (r *Reconciler) isDeleteNodeAllowed(ctx context.Context, cluster *clusterv1
 	return nil
 }
 
+// isVersionAllowed returns true if the machine version conforms to the Kubernetes version skew policy.
+// If the Machine is part of the control plane, then its major.minor version must be greater than or equal to the
+// control plane actual major.minor version, and less than or equal to the control plane desired major.minor version.
+// If the Machine is not part of the control plane, then its major.minor version must be less than or equal to the control
+// plane actual major.minor version.
+func (r *Reconciler) isVersionAllowed(ctx context.Context, machine *clusterv1.Machine, cluster *clusterv1.Cluster) (bool, error) {
+	// Skip the check if Version nil, because it is an optional field.
+	if machine.Spec.Version == nil {
+		return true, nil
+	}
+
+	// Skip the check if ControlPlaneRef is nil, because it is an optional field.
+	if cluster.Spec.ControlPlaneRef == nil {
+		return true, nil
+	}
+
+	mv, err := semver.ParseTolerant(*machine.Spec.Version)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse machine version")
+	}
+
+	controlPlane, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Spec.ControlPlaneRef.Namespace)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get control plane object")
+	}
+
+	controlPlaneActualVersionStr, err := contract.ControlPlane().StatusVersion().Get(controlPlane)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to read control plane actual version")
+	}
+	// controlPlaneVersionStr is not nil, because Get did not return an error.
+	controlPlaneActualVersion, err := semver.ParseTolerant(*controlPlaneActualVersionStr)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse control plane actual version")
+	}
```

Comment:
`status.version` was an optional field that was added later, not sure if it's in the current contract (cc @fabriziopandini), although `spec.version` should be there. If the current version in status is not available, can we fallback to only look at the desired one instead?

Reply:
If only .spec.version exists and .status.version doesn't, in most cases this means that the first control plane machine isn't up yet. In the topology reconcile we're interpreting this as "control plane is provisioning"

Reply:
Thanks for the additional context. I have a TODO to handle the absence of Status.Version.